### PR TITLE
(INT): Add constraint to where clause intention

### PR DESCRIPTION
**Pull request by alobb**
This implements (part of) the functionality for #694.  It adds a new intention that moves type constraints to a where clause for functions (currently only works on functions).  I've never written Kotlin before or worked with custom language plugins for Intellij, so let me know if anything looks wrong/un-idiomatic!  Also, would it make sense to get this working with structs/traits/enums on this PR, or a separate one?

### CODE REVIEW

**Comment by alobb** [comment]
Something I noticed in some (but not all) of the other inspections is that they reformat the code after inserting a PSI Element; should I do that here as well? What determines when this should be done?

File: src/main/kotlin/org/rust/ide/intentions/MoveTypeConstraintToWhereClauseIntention.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+class MoveTypeConstraintToWhereClauseIntention : PsiElementBaseIntentionAction() {
+    override fun getText() = "Move type constraint to where clause"
+    override fun getFamilyName() = text
+    override fun startInWriteAction() = true
+
+    override fun isAvailable(project: Project, editor: Editor, element: PsiElement): Boolean {
+        if (!element.isWritable) return false
+
+        val function = element.parentOfType<RustFnItemElement>() ?: return false
+        val genericTypeBounds = function.genericParams?.typeParamList?.filter { it.typeParamBounds != null }.orEmpty()
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/intentions/MoveTypeConstraintToWhereClauseIntention.kt]: There's `filterNotNull`.

File: src/main/kotlin/org/rust/lang/core/psi/RustElementFactory.kt [hunk 2]
```diff
@@ -42,6 +42,17 @@ object RustElementFactory {
     fun createIdentifier(project: Project, name: String): PsiElement =
         createFromText<RustModDeclItemElement>(project, "mod $name;")!!.identifier
 
+    fun createWhereClause(project: Project, lifetimeBounds: List<RustLifetimeParamElement>, typeBounds: List<RustTypeParamElement>): RustWhereClauseElement? {
+        val lifetimes = lifetimeBounds
+            .filter { it.lifetimeParamBounds != null }
+            .mapNotNull { it.node.text }
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/psi/RustElementFactory.kt]: I believe you can use a single `.mapNotNull { it.lifetimeParamBouds?.node?.text }` here.

**Comment by matklad** [comment]
>What determines when this should be done?

It's always safe to reformat code after refactorings, although if changes are local (like in this case) you can get away without reformatting.

**Comment by matklad** [comment]
Thanks!

**Pull request merged**